File: deploy/new_html/services/canvasService.ts
import { apiJson } from './httpClient';

export interface CanvasBoardPayload {
  name?: string;
  description?: string;
}

export interface CanvasNodePayload {
  [key: string]: any;
}

export async function createCanvasBoard(projectId: string, name = '未命名画布', description = '') {
  return apiJson<any>('/api/canvas/boards', {
    method: 'POST',
    body: JSON.stringify({ project_id: projectId, name, description }),
  }, 'createCanvasBoard');
}

export async function getCanvasBoards(projectId: string) {
  const qs = new URLSearchParams({ project_id: projectId });
  return apiJson<any>(`/api/canvas/boards?${qs.toString()}`, { method: 'GET' }, 'getCanvasBoards');
}

export async function getCanvasBoardDetail(boardId: string) {
  return apiJson<any>(`/api/canvas/boards/${boardId}`, { method: 'GET' }, 'getCanvasBoardDetail');
}

export async function updateCanvasBoard(boardId: string, data: CanvasBoardPayload) {
  return apiJson<any>(`/api/canvas/boards/${boardId}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  }, 'updateCanvasBoard');
}

export async function deleteCanvasBoard(boardId: string) {
  return apiJson<any>(`/api/canvas/boards/${boardId}`, { method: 'DELETE' }, 'deleteCanvasBoard');
}

export async function createCanvasNode(boardId: string, nodeType: string, x = 0, y = 0, data?: CanvasNodePayload) {
  return apiJson<any>('/api/canvas/nodes', {
    method: 'POST',
    body: JSON.stringify({ board_id: boardId, node_type: nodeType, x, y, data }),
  }, 'createCanvasNode');
}

export async function updateCanvasNode(nodeId: string, data: CanvasNodePayload) {
  return apiJson<any>(`/api/canvas/nodes/${nodeId}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  }, 'updateCanvasNode');
}

export async function deleteCanvasNode(nodeId: string) {
  return apiJson<any>(`/api/canvas/nodes/${nodeId}`, { method: 'DELETE' }, 'deleteCanvasNode');
}

export async function createCanvasConnection(boardId: string, sourceNodeId: string, targetNodeId: string) {
  return apiJson<any>('/api/canvas/connections', {
    method: 'POST',
    body: JSON.stringify({ board_id: boardId, source_node_id: sourceNodeId, target_node_id: targetNodeId }),
  }, 'createCanvasConnection');
}

export async function deleteCanvasConnection(connectionId: string) {
  return apiJson<any>(`/api/canvas/connections/${connectionId}`, { method: 'DELETE' }, 'deleteCanvasConnection');
}

File: deploy/services/canvas_service.py
"""Canvas board, node, and connection business logic."""
from __future__ import annotations

from typing import Any, Dict, Optional


class CanvasServiceError(RuntimeError):
    pass


class CanvasPermissionDenied(CanvasServiceError):
    pass


class CanvasBoardNotFound(CanvasServiceError):
    pass


async def _require_project_permission(
    project_id: str,
    user_id: str,
    required_role: str,
    *,
    project_member_dao: Any,
) -> None:
    has_permission = await project_member_dao.check_permission(project_id, user_id, required_role)
    if not has_permission:
        raise CanvasPermissionDenied(f"Project permission required: {required_role}")


async def _get_board_or_raise(board_id: str, *, canvas_board_dao: Any) -> Dict[str, Any]:
    board = await canvas_board_dao.get_board(board_id)
    if not board:
        raise CanvasBoardNotFound("Canvas board not found")
    return board


async def create_canvas_board(
    *,
    project_id: str,
    user_id: str,
    name: Optional[str],
    description: Optional[str],
    project_member_dao: Any,
    canvas_board_dao: Any,
) -> Dict[str, Any]:
    await _require_project_permission(
        project_id,
        user_id,
        "member",
        project_member_dao=project_member_dao,
    )
    board = await canvas_board_dao.create_board(project_id, user_id, name, description)
    return {"success": True, "board": board}


async def list_canvas_boards(
    *,
    project_id: str,
    user_id: str,
    project_member_dao: Any,
    canvas_board_dao: Any,
) -> Dict[str, Any]:
    await _require_project_permission(
        project_id,
        user_id,
        "readonly",
        project_member_dao=project_member_dao,
    )
    boards = await canvas_board_dao.get_project_boards(project_id)
    return {"success": True, "boards": boards}


async def get_canvas_board_detail(
    *,
    board_id: str,
    user_id: str,
    project_member_dao: Any,
    canvas_board_dao: Any,
    canvas_node_dao: Any,
    canvas_connection_dao: Any,
) -> Dict[str, Any]:
    board = await _get_board_or_raise(board_id, canvas_board_dao=canvas_board_dao)
    await _require_project_permission(
        board["project_id"],
        user_id,
        "readonly",
        project_member_dao=project_member_dao,
    )
    nodes = await canvas_node_dao.get_board_nodes(board_id)
    connections = await canvas_connection_dao.get_board_connections(board_id)
    return {
        "success": True,
        "board": board,
        "nodes": nodes,
        "connections": connections,
    }


async def update_canvas_board(
    *,
    board_id: str,
    user_id: str,
    fields: Dict[str, Any],
    project_member_dao: Any,
    canvas_board_dao: Any,
) -> Dict[str, Any]:
    board = await _get_board_or_raise(board_id, canvas_board_dao=canvas_board_dao)
    await _require_project_permission(
        board["project_id"],
        user_id,
        "member",
        project_member_dao=project_member_dao,
    )
    await canvas_board_dao.update_board(board_id, **fields)
    return {"success": True}


async def delete_canvas_board(
    *,
    board_id: str,
    user_id: str,
    project_member_dao: Any,
    canvas_board_dao: Any,
) -> Dict[str, Any]:
    board = await _get_board_or_raise(board_id, canvas_board_dao=canvas_board_dao)
    await _require_project_permission(
        board["project_id"],
        user_id,
        "admin",
        project_member_dao=project_member_dao,
    )
    await canvas_board_dao.delete_board(board_id)
    return {"success": True}


async def create_canvas_node(
    *,
    board_id: str,
    node_type: str,
    x: Optional[float],
    y: Optional[float],
    width: Optional[float],
    height: Optional[float],
    data: Optional[dict],
    user_id: str,
    project_member_dao: Any,
    canvas_board_dao: Any,
    canvas_node_dao: Any,
) -> Dict[str, Any]:
    board = await _get_board_or_raise(board_id, canvas_board_dao=canvas_board_dao)
    await _require_project_permission(
        board["project_id"],
        user_id,
        "member",
        project_member_dao=project_member_dao,
    )
    node = await canvas_node_dao.create_node(board_id, node_type, x, y, width, height, data)
    return {"success": True, "node": node}


async def update_canvas_node(
    *,
    node_id: str,
    fields: Dict[str, Any],
    canvas_node_dao: Any,
) -> Dict[str, Any]:
    await canvas_node_dao.update_node(node_id, **fields)
    return {"success": True}


async def delete_canvas_node(
    *,
    node_id: str,
    canvas_node_dao: Any,
) -> Dict[str, Any]:
    await canvas_node_dao.delete_node(node_id)
    return {"success": True}


async def create_canvas_connection(
    *,
    board_id: str,
    source_node_id: str,
    target_node_id: str,
    source_port: Optional[str],
    target_port: Optional[str],
    label: Optional[str],
    canvas_connection_dao: Any,
) -> Dict[str, Any]:
    canvas_link = await canvas_connection_dao.create_connection(
        board_id,
        source_node_id,
        target_node_id,
        source_port,
        target_port,
        label,
    )
    return {"success": True, "connection": canvas_link}


async def delete_canvas_connection(
    *,
    connection_id: str,
    canvas_connection_dao: Any,
) -> Dict[str, Any]:
    await canvas_connection_dao.delete_connection(connection_id)
    return {"success": True}
